dvc/remote/base.py

```diff
@@ -381,18 +382,105 @@ def _save(self, path_info, checksum):
+                if to_info.scheme != self.scheme:
+                    raise NotImplementedError
+
+                if from_info.scheme != "local":
+                    raise NotImplementedError
+
+                msg = "Uploading '{}' to '{}'"
+                logger.debug(msg.format(from_info, to_info))
+
+                if not name:
+                    name = from_info.name
+
+                if not no_progress_bar:
+                    progress.update_target(name, 0, None)
+
+                try:
+                    self._upload(
+                        from_info.fspath,
+                        to_info,
+                        name=name,
+                        ctx=ctx,
+                        no_progress_bar=no_progress_bar,
+                    )
+                except Exception:
+                    msg = "failed to upload '{}' to '{}'"
+                    logger.exception(msg.format(from_info, to_info))
+                    continue
+
+                if not no_progress_bar:
+                    progress.finish_target(name)
+
     def download(
         self,
         from_infos,
         to_infos,
         no_progress_bar=False,
-        name=None,
+        names=None,
         resume=False,
     ):
-        raise RemoteActionNotImplemented("download", self.scheme)
+        if not hasattr(self, "_download"):
+            raise RemoteActionNotImplemented("download", self.scheme)
 
-    def upload(self, from_infos, to_infos, names=None, no_progress_bar=False):
-        raise RemoteActionNotImplemented("upload", self.scheme)
+        names = self._verify_path_args(from_infos, to_infos, names)
+
+        with self.transfer_context() as ctx:
+            for to_info, from_info, name in zip(to_infos, from_infos, names):
+                if from_info.scheme != self.scheme:
+                    raise NotImplementedError
+
+                if to_info.scheme == self.scheme != "local":
+                    self.copy(from_info, to_info, ctx=ctx)
+                    continue
+
+                if to_info.scheme != "local":
+                    raise NotImplementedError
+
+                msg = "Downloading '{}' to '{}'".format(from_info, to_info)
+                logger.debug(msg)
+
+                tmp_file = tmp_fname(to_info)
+                if not name:
+                    name = to_info.name
+
+                if not no_progress_bar:
+                    # real progress is not always available,
+                    # lets at least show start and finish
+                    progress.update_target(name, 0, None)
+
+                makedirs(fspath_py35(to_info.parent), exist_ok=True)
+
+                try:
+                    self._download(
+                        from_info,
+                        tmp_file,
+                        name=name,
+                        ctx=ctx,
+                        resume=resume,
+                        no_progress_bar=no_progress_bar,
+                    )
+                except Exception:
```

Comment:
don't catch Exception, or raise it - failing fast is better if we don't know for sure what's happening

Reply:
This is an old behavior, I don't change it. This refactor is non-semantic one, aside from fixing some bugs. I am just trying to minimize a number of places I will need to apply fix for #2038 to.

The original idea as I see it was to download/upload as much as we can even in presence of some errors.

Reply:
yep, I saw it now that it was before